src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.cs

```diff
@@ -507,7 +472,9 @@ public string ToUpper(string str)
                 return InvariantModeCasing.ToUpper(str);
             }
 
-            return ChangeCaseCommon<ToUpperConversion>(str);
+            string result = string.FastAllocateString(str.Length);
```

Comment:
This is going to regress allocations for cases when the string lower-cased already. The existing code just returns the existing string for those cases.

Reply:
Good point, I reverted the change for string, now it's only for Spans as I don't have a good solution for this problem, I made a quick prototype and it was too big (I had to duplicate code for spans which don't care about this problem) + overhead from an additional branch inside the loop + I had to keep the old scalar path.

Reply:
Will re-run the benchmarks for the Span